File: projects/src/main/java/io/nirahtech/erp/projects/Milestone.java
package io.nirahtech.erp.projects;

import java.io.File;
import java.io.Serializable;
import java.time.Duration;
import java.time.LocalDate;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

public final class Milestone implements Serializable {
	private final MilestoneIdentifier id;
	private final String name;
	private Duration duration;
	private String description = null;
	private ProjectMember referent = null;
	private LocalDate startDate = LocalDate.now();
	private Status status = Status.IDLE;
	private final Set<File> documentations = new HashSet<>();
	private final Set<String> labels = new HashSet<>();
	private final Map<Status, Runnable> eventListeners = new HashMap<>();

	public Milestone(
			final MilestoneIdentifier id,
			final String name,
			final String description,
			final ProjectMember referent,
			final LocalDate startDate,
			final Duration duration,
			final Status status,
			final Set<File> documentations,
			final Set<String> labels) {
		this.id = id;
		this.name = name;
		this.description = description;
		this.referent = referent;
		this.startDate = startDate;
		this.duration = duration;
		this.status = status;
		this.documentations.addAll(documentations);
		this.labels.addAll(labels);
	}

	public String getDescription() {
		return description;
	}
	public void setDescription(String description) {
		this.description = description;
	}
	public void setReferent(ProjectMember referent) {
		this.referent = referent;
	}
	public void setStartDate(LocalDate startDate) {
		this.startDate = startDate;
	}
	public void setStatus(Status status) {
		this.status = status;
		if (this.eventListeners.containsKey(status)) {
			Runnable callback = this.eventListeners.get(status);
			callback.run();
		}
	}
	public void addLabel(final String label) {
		this.labels.add(label);
	}
	public void addLabels(final String... labels) {
		this.labels.addAll(Set.of(labels));
	}
	public void addDocumentation(final File documentation) {
		this.documentations.add(documentation);
	}
	public void addDocumentations(final File... documentations) {
		this.documentations.addAll(Set.of(documentations));
	}
	public Collection<File> getDocumentations() {
		return documentations;
	}

	public Duration getDuration() {
		return duration;
	}

	public MilestoneIdentifier getId() {
		return id;
	}

	public Collection<String> getLabels() {
		return labels;
	}

	public String getName() {
		return name;
	}

	public ProjectMember getReferent() {
		return referent;
	}

	public LocalDate getStartDate() {
		return startDate;
	}

	public Status getStatus() {
		return status;
	}

	public LocalDate getDeadLine() {
		return this.startDate.plus(this.duration);
	}
	public void setDuration(Duration duration) {
		this.duration = duration;
	}

	public Optional<LocalDate> getEstimatedEndDate() {
		Optional<LocalDate> estimatedEndDate = Optional.empty();
		
		return estimatedEndDate;
	}

	public void addEventListenerOnStatusChanged(final Status status, final Runnable callback) {
		this.eventListeners.put(status, callback);
	}
}

File: webapp/src/main/react/src/data/Milestone.ts
import Duration from "./Duration";
import LocalDate from "./LocalDate";
import { MilestoneIdentifier } from "./MilestoneIdentifier";
import ProjectMember from "./ProjectMember";
import Status from "./Status";

class Milestone {
  private readonly id: MilestoneIdentifier;
  private readonly name: string;
  private description: string | null = null;
  private referent: ProjectMember | null = null;
  private startDate: LocalDate = LocalDate.now();
  private status: Status = Status.IDLE;
  private duration: Duration;
  private readonly documentations: Set<File> = new Set<File>();
  private readonly labels: Set<string> = new Set<string>();
  private readonly eventListeners: Map<Status, Function> = new Map<Status, Function>();

  constructor(
    id: MilestoneIdentifier,
    name: string,
    description: string | null,
    referent: ProjectMember | null,
    startDate: LocalDate,
    duration: Duration,
    status: Status,
    documentations: Set<File>,
    labels: Set<string>
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.referent = referent;
    this.startDate = startDate;
    this.duration = duration;
    this.status = status;
    this.documentations = new Set<File>(documentations);
    this.labels = new Set<string>(labels);
    this.eventListeners = new Map<Status, Function>();
  }

  getDescription(): string | null {
    return this.description;
  }

  setDescription(description: string | null): void {
    this.description = description;
  }

  setReferent(referent: ProjectMember | null): void {
    this.referent = referent;
  }

  setStartDate(startDate: LocalDate): void {
    this.startDate = startDate;
  }

  setStatus(status: Status): void {
    this.status = status;
    if (this.eventListeners.has(status)) {
      const callback = this.eventListeners.get(status);
      if (callback) {
        callback();
      }
    }
  }

  addLabel(label: string): void {
    this.labels.add(label);
  }

  addLabels(labels: string[]): void {
    labels.forEach((label) => {
      this.labels.add(label);
    });
  }

  addDocumentation(documentation: File): void {
    this.documentations.add(documentation);
  }

  addDocumentations(documentations: File[]): void {
    documentations.forEach((documentation) => {
      this.documentations.add(documentation);
    });
  }

  getDocumentations(): Set<File> {
    return this.documentations;
  }

  getDuration(): Duration {
    return this.duration;
  }

  getId(): MilestoneIdentifier {
    return this.id;
  }

  getLabels(): Set<string> {
    return this.labels;
  }

  getName(): string {
    return this.name;
  }

  getReferent(): ProjectMember | null {
    return this.referent;
  }

  getStartDate(): LocalDate {
    return this.startDate;
  }

  getStatus(): Status {
    return this.status;
  }

  getDeadline(): LocalDate {
    return this.startDate.plus(this.duration);
  }

  setDuration(duration: Duration): void {
    this.duration = duration;
  }

  getEstimatedEndDate(): LocalDate | undefined {
    return this.startDate.plus(this.duration);
  }

  addEventListenerOnStatusChanged(status: Status, callback: Function): void {
    this.eventListeners.set(status, callback);
  }
}

export { Milestone };
